src/Playwright/Transport/WebSocketTransport.cs

```diff
@@ -0,0 +1,183 @@
+    {
+        private const int DefaultBufferSize = 16000;  // Byte buffer size
+        private readonly ClientWebSocket _webSocket;
+        private readonly string _wsEndpoint;
+        private readonly CancellationTokenSource _readerCancellationSource;
+        private readonly BrowserTypeConnectOptions _options;
+        private readonly int _slowMo;
+        private readonly int _timeout;
+
+        internal WebSocketTransport(
+            string wsEndpoint = default,
+            BrowserTypeConnectOptions options = default)
+        {
+            _webSocket = new ClientWebSocket();
+            _wsEndpoint = wsEndpoint;
+            _options = options;
+            _slowMo = _options?.SlowMo != null ? (int)_options?.SlowMo : 0;
+            _timeout = _options?.Timeout != null ? (int)_options?.Timeout : 30000;
+            _readerCancellationSource = new CancellationTokenSource(_timeout);
+            SetRequestHeaders();
+            _ = ConnectAsync();
+        }
+
+        /// <inheritdoc cref="IDisposable.Dispose"/>
+        ~WebSocketTransport() => Dispose(false);
+
+        public event EventHandler<MessageReceivedEventArgs> MessageReceived;
+
+        public event EventHandler<LogReceivedEventArgs> LogReceived;
+
+        public event EventHandler<TransportClosedEventArgs> TransportClosed;
+
+        public bool IsClosed { get; private set; }
+
+        public async Task SendAsync(string message)
+        {
+            await Task.Delay(_slowMo).ConfigureAwait(false);
+
+            try
+            {
+                var messageBuffer = Encoding.UTF8.GetBytes(message);
+                if (!_readerCancellationSource.IsCancellationRequested)
+                {
+                    await _webSocket.SendAsync(new ArraySegment<byte>(messageBuffer, 0, messageBuffer.Length), WebSocketMessageType.Text, true, _readerCancellationSource.Token).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                Close(ex);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Close(string closeReason)
+        {
+            if (!IsClosed)
+            {
+                IsClosed = true;
+                HandleSocketClose(closeReason);
+                TransportClosed?.Invoke(this, new TransportClosedEventArgs { CloseReason = closeReason });
+                _readerCancellationSource?.Cancel();
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Close(Exception ex)
+        {
+            System.Diagnostics.Debug.WriteLine(ex);
+            Close(ex.ToString());
+        }
+
+        private void HandleSocketClose(string closeReason)
+        {
+            _webSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, closeReason, _readerCancellationSource.Token).ConfigureAwait(false);
```

Comment:
It is still weird that you react to the "socket closed" event via closing the socket. What should happen here instead is handling the socket closure: proper tear down and emitting the transport closed event. And it should happen when the socket closes, either by remote or by local.

Reply:
updated, it will invoke TransportClosed event

Reply:
fixed